grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerInterceptor.java

```diff
@@ -0,0 +1,137 @@
+import io.grpc.ClientCall;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerConfigProvider;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerInterceptor;
+
+@Slf4j
+@Singleton
+public class ResilienceCircuitBreakerInterceptor extends CircuitBreakerInterceptor {
+
+  public static final CallOptions.Key<String> CIRCUIT_BREAKER_KEY =
+      CallOptions.Key.createWithDefault("circuitBreakerKey", "default");
+  private final CircuitBreakerRegistry resilicenceCircuitBreakerRegistry;
+  private final CircuitBreakerConfigProvider circuitBreakerConfigProvider;
+  private final Map<String, CircuitBreakerConfig> resilienceCircuitBreakerConfig;
+  private final ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider;
+  private final Clock clock;
+
+  public ResilienceCircuitBreakerInterceptor(Config config, Clock clock) {
+    this.circuitBreakerConfigProvider = new CircuitBreakerConfigProvider(config);
+    this.resilienceCircuitBreakerConfig =
+        ResilienceCircuitBreakerConfigParser.getCircuitBreakerConfigs(
+            circuitBreakerConfigProvider.getConfigMap());
+    this.resilicenceCircuitBreakerRegistry =
+        new ResilienceCircuitBreakerRegistryProvider(resilienceCircuitBreakerConfig)
+            .getCircuitBreakerRegistry();
+    this.resilienceCircuitBreakerProvider =
+        new ResilienceCircuitBreakerProvider(
+            resilicenceCircuitBreakerRegistry, resilienceCircuitBreakerConfig);
+    this.clock = clock;
+  }
+
+  @VisibleForTesting
+  public ResilienceCircuitBreakerInterceptor(
+      Config config,
+      Clock clock,
+      CircuitBreakerRegistry resilicenceCircuitBreakerRegistry,
+      ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider) {
+    this.circuitBreakerConfigProvider = new CircuitBreakerConfigProvider(config);
+    this.resilienceCircuitBreakerConfig =
+        ResilienceCircuitBreakerConfigParser.getCircuitBreakerConfigs(
+            circuitBreakerConfigProvider.getConfigMap());
+    this.resilicenceCircuitBreakerRegistry = resilicenceCircuitBreakerRegistry;
+    this.resilienceCircuitBreakerProvider = resilienceCircuitBreakerProvider;
+    this.clock = clock;
+  }
+
+  @Override
+  protected boolean isCircuitBreakerEnabled() {
+    return circuitBreakerConfigProvider.isCircuitBreakerEnabled();
+  }
+
+  @Override
+  protected <ReqT, RespT> ClientCall<ReqT, RespT> createInterceptedCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    // Get circuit breaker key from CallOptions
+    String circuitBreakerKey = callOptions.getOption(CIRCUIT_BREAKER_KEY);
```

Comment:
Let's discuss this. I think using a global interceptor with magic attributes on the call is way more confusing and easy to mess up than a contract like below, where we attach the circuit breaker for our stub to our stub.


```
CircuitBreakerConfig parsedConfig = CircuitBreakerConfigParser.parse(genericConfig); // Or use the builder and roll your own
RequestInterceptor interceptor = new CircuitBreakerRequestInterceptor(parsedConfig, clock);

myStub = myStub.withInterceptor(interceptor);
// Now all calls via myStub go through the circuit breaker
```

Reply:
Discussed offline and updated the patch